File: go/solution/0006.go
/*
 * @Author: whe
 * @Date: 2022-03-01 10:44:03
 */
package solution

func convert(s string, numRows int) string {
	buffer := []byte{}
	if numRows == 1 {
		return s
	}
	for i := 0; i < numRows; i++ {
		//z字两端直接写入
		if i == 0 || i == numRows-1 {
			for j := i; j < len(s); j += 2*numRows - 2 {
				buffer = append(buffer, s[j])
			}
		} else {
			// z字中间同一行取模后相加为固定值
			for j := i; j < len(s); j += 2*numRows - 2 {
				buffer = append(buffer, s[j])
				n := j + 2*numRows - 2 - j%(2*numRows-2)*2
				if n < len(s) {
					buffer = append(buffer, s[n])
				}
			}
		}
	}
	return string(buffer)
}

File: java/src/_0006.java
/*
 * @Author: whe
 * @Date: 2022-03-01 10:49:48
 */
public class _0006 {
    public String convert(String s, int numRows) {
        StringBuffer ans = new StringBuffer();
        char[] sArray = s.toCharArray();
        if (numRows == 1) {
            return s;
        }
        for (int i = 0; i < numRows; i++) {
            if (i == 0 || i == numRows - 1) {
                for (int j = i; j < s.length(); j += 2 * (numRows - 1)) {
                    ans.append(sArray[j]);
                }
            } else {
                for (int j = i; j < s.length(); j += 2 * (numRows - 1)) {
                    ans.append(sArray[j]);
                    int n = j + 2 * numRows - 2 - j % (2 * numRows - 2) * 2;
                    if (n < s.length()) {
                        ans.append(sArray[n]);
                    }
                }
            }
        }
        return ans.toString();
    }
}
